third_party/terraform/utils/iam_storage_bucket.go

```diff
@@ -0,0 +1,175 @@
+	"bucket": {
+		Type:             schema.TypeString,
+		Required:         true,
+		ForceNew:         true,
+		DiffSuppressFunc: compareSelfLinkOrResourceName,
+	},
+}
+
+type StorageBucketIamUpdater struct {
+	bucket string
+	d      *schema.ResourceData
+	Config *Config
+}
+
+func StorageBucketIamUpdaterProducer(d *schema.ResourceData, config *Config) (ResourceIamUpdater, error) {
+	values := make(map[string]string)
+
+	if v, ok := d.GetOk("bucket"); ok {
+		values["bucket"] = v.(string)
+	}
+
+	// We may have gotten either a long or short name, so attempt to parse long name if possible
+	m, err := getImportIdQualifiers([]string{"b/(?P<bucket>[^/]+)", "(?P<bucket>[^/]+)"}, d, config, d.Get("bucket").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range m {
+		values[k] = v
+	}
+
+	u := &StorageBucketIamUpdater{
+		bucket: values["bucket"],
+		d:      d,
+		Config: config,
+	}
+
+	d.Set("bucket", u.GetResourceId())
+
+	return u, nil
+}
+
+func StorageBucketIdParseFunc(d *schema.ResourceData, config *Config) error {
+	values := make(map[string]string)
+
+	m, err := getImportIdQualifiers([]string{"b/(?P<bucket>[^/]+)", "(?P<bucket>[^/]+)"}, d, config, d.Id())
+	if err != nil {
+		return err
+	}
+
+	for k, v := range m {
+		values[k] = v
+	}
+
+	u := &StorageBucketIamUpdater{
+		bucket: values["bucket"],
+		d:      d,
+		Config: config,
+	}
+	d.Set("bucket", u.GetResourceId())
+	d.SetId(u.GetResourceId())
+	return nil
+}
+
+func (u *StorageBucketIamUpdater) GetResourceIamPolicy() (*cloudresourcemanager.Policy, error) {
+	url, err := u.qualifyBucketUrl("iam")
+	if err != nil {
+		return nil, err
+	}
+
+	var obj map[string]interface{}
+	url, err = addQueryParams(url, map[string]string{"optionsRequestedPolicyVersion": fmt.Sprintf("%d", iamPolicyVersion)})
+	if err != nil {
+		return nil, err
+	}
+
+	policy, err := sendRequest(u.Config, "GET", "", url, obj)
+	if err != nil {
+		return nil, errwrap.Wrapf(fmt.Sprintf("Error retrieving IAM policy for %s: {{err}}", u.DescribeResource()), err)
+	}
+
+	out := &cloudresourcemanager.Policy{}
+	err = Convert(policy, out)
+	if err != nil {
+		return nil, errwrap.Wrapf("Cannot convert a policy to a resource manager policy: {{err}}", err)
+	}
+
+	return out, nil
+}
+
+func (u *StorageBucketIamUpdater) SetResourceIamPolicy(policy *cloudresourcemanager.Policy) error {
+	url, err := u.qualifyBucketUrl("iam")
+	if err != nil {
+		return err
+	}
+	err = u.SendPolicyRequest(url, policy)
+	if err != nil {
+		// Cloud storage guards against race conditions with precondition checks, the policy
+		// may be updated by the time we need to delete. See https://cloud.google.com/storage/docs/generations-preconditions
+		gerr, ok := err.(*googleapi.Error)
+		if ok && gerr.Code == 412 {
+			log.Printf("[DEBUG] Received 412 error, retrying once with updated policy")
+			updatedPolicy, err := u.GetResourceIamPolicy()
+			if err != nil {
+				return err
+			}
+			err = u.SendPolicyRequest(url, updatedPolicy)
```

Comment:
What is the point of this? If we get the 412 error then we retrieve the current policy and set the policy to be what we got? It seems like that would be a no-op

Reply:
The issue comes from a race condition where the policy has been updated by the time the DELETE request is sent. This is a refresh+retry instead of a regular retry.

Reply:
You are right, for some reason I though the policy was getting modified in the SendPolicyRequest function. The only way this is working then is if the policy is already removed by the time this SetResourceIamPolicy function is called. So essentially my change would catching the 412, retrieving the empty policy, and sending it back (only explanation on how this fixes the error). I'll double check